### ISSUE TITLE
Fail to start docker containers in WSL2 with 1.12.1

### ISSUE THREAD

**Issue by itmo153277** [NVIDIA/nvidia-container-toolkit#53]
Docker containers fail to start in WSL2 with CUDA images.

Docker freezes during start. When I kill `nvidia-container-cli` it reports segfault.

The issue occurs with 1.12.1+. There is no such problem in 1.12.0.

<details>
<summary>Example logs</summary>

```
$ cat /proc/version
Linux version 5.15.90.1-microsoft-standard-WSL2 (oe-user@oe-host) (x86_64-msft-linux-gcc (GCC) 9.3.0, GNU ld (GNU Binutils) 2.34.0.20200220) #1 SMP Fri Jan 27 02:56:13 UTC 2023

$ docker --version
Docker version 23.0.1, build a5ee5b1

$ nvidia-container-cli info
NVRM version:   531.18
CUDA version:   12.1

Device Index:   0
Device Minor:   0
Model:          NVIDIA GeForce RTX 3060 Laptop GPU
Brand:          GeForce
GPU UUID:       GPU-1efeaa36-8482-25de-66dc-65f810bca2b8
Bus Location:   00000000:01:00.0
Architecture:   8.6

$ nvidia-container-cli --version
cli-version: 1.12.1
lib-version: 1.12.1
build date: 2023-03-10T11:34+00:00
build revision: 7440a1ead8e4fe35edf1c973c73a662108b21a1f
build compiler: x86_64-linux-gnu-gcc-7 7.5.0
build platform: x86_64
build flags: -D_GNU_SOURCE -D_FORTIFY_SOURCE=2 -DNDEBUG -std=gnu11 -O2 -g -fdata-sections -ffunction-sections -fplan9-extensions -fstack-protector -fno-strict-aliasing -fvisibility=hidden -Wall -Wextra -Wcast-align -Wpointer-arith -Wmissing-prototypes -Wnonnull -Wwrite-strings -Wlogical-op -Wformat=2 -Wmissing-format-attribute -Winit-self -Wshadow -Wstrict-prototypes -Wunreachable-code -Wconversion -Wsign-conversion -Wno-unknown-warning-option -Wno-format-extra-args -Wno-gnu-alignof-expression -Wl,-zrelro -Wl,-znow -Wl,-zdefs -Wl,--gc-sections

$ docker run --gpus all nvcr.io/nvidia/k8s/cuda-sample:nbody nbody -gpu -benchmark
<...stuck here until I manually kill nvidia-container-cli...>
docker: Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: signal: segmentation fault, stdout: , stderr: Auto-detected mode as 'legacy': unknown.
ERRO[0027] error waiting for container:
```

```
$ nvidia-container-cli --version
cli-version: 1.12.0
lib-version: 1.12.0
build date: 2023-03-14T06:55+00:00
build revision: 7678e1af094d865441d0bc1b97c3e72d15fcab50
build compiler: x86_64-linux-gnu-gcc-7 7.5.0
build platform: x86_64
build flags: -D_GNU_SOURCE -D_FORTIFY_SOURCE=2 -DNDEBUG -std=gnu11 -O2 -g -fdata-sections -ffunction-sections -fplan9-extensions -fstack-protector -fno-strict-aliasing -fvisibility=hidden -Wall -Wextra -Wcast-align -Wpointer-arith -Wmissing-prototypes -Wnonnull -Wwrite-strings -Wlogical-op -Wformat=2 -Wmissing-format-attribute -Winit-self -Wshadow -Wstrict-prototypes -Wunreachable-code -Wconversion -Wsign-conversion -Wno-unknown-warning-option -Wno-format-extra-args -Wno-gnu-alignof-expression -Wl,-zrelro -Wl,-znow -Wl,-zdefs -Wl,--gc-sections

$ docker run --gpus all nvcr.io/nvidia/k8s/cuda-sample:nbody nbody -gpu -benchmark
Run "nbody -benchmark [-numbodies=<numBodies>]" to measure performance.
        -fullscreen       (run n-body simulation in fullscreen mode)
        -fp64             (use double precision floating point values for simulation)
        -hostmem          (stores simulation data in host memory)
        -benchmark        (run benchmark to measure performance)
        -numbodies=<N>    (number of bodies (>= 1) to run in simulation)
        -device=<d>       (where d=0,1,2.... for the CUDA device to use)
        -numdevices=<i>   (where i=(number of CUDA devices > 0) to use for simulation)
        -compare          (compares simulation results running once on the default GPU and once on the CPU)
        -cpu              (run n-body simulation on the CPU)
        -tipsy=<file.bin> (load a tipsy model file for simulation)

NOTE: The CUDA Samples are not meant for performance measurements. Results may vary when GPU Boost is enabled.

> Windowed mode
> Simulation data stored in video memory
> Single precision floating point simulation
> 1 Devices used for simulation
GPU Device 0: "Ampere" with compute capability 8.6

> Compute 8.6 CUDA device: [NVIDIA GeForce RTX 3060 Laptop GPU]
30720 bodies, total time for 10 iterations: 31.060 ms
= 303.838 billion interactions per second
= 6076.750 single-precision GFLOP/s at 20 flops per interaction
```

</details>

**Comment by elezar** [comment]
Do the `nvidia-container-cli list` or `nvidia-container-cli info` commands work as expected?

**Comment by elezar** [comment]
@klueska instead of `return 0` we not `goto success` on WSL2 systems. This triggers a free of a `NULL`. Would this be causing a segfault?

@itmo153277 would you be able to test a build off `HEAD` of the `release-1.12` branch?

**Comment by itmo153277** [comment]
@elezar 
Yes, both `list` and `info` work as expected on 1.12.1

```
$ nvidia-container-cli info
NVRM version:   531.18
CUDA version:   12.1

Device Index:   0
Device Minor:   0
Model:          NVIDIA GeForce RTX 3060 Laptop GPU
Brand:          GeForce
GPU UUID:       GPU-1efeaa36-8482-25de-66dc-65f810bca2b8
Bus Location:   00000000:01:00.0
Architecture:   8.6

$ nvidia-container-cli list
/dev/dxg
/usr/lib/wsl/drivers/nvlti.inf_amd64_b58fa54b9732141a/nvidia-smi
/usr/lib/wsl/lib/libnvidia-ml.so.1
/usr/lib/wsl/lib/libcuda.so.1
/usr/lib/wsl/lib/libnvidia-encode.so.1
/usr/lib/wsl/lib/libnvidia-opticalflow.so.1
/usr/lib/wsl/lib/libnvcuvid.so.1
/usr/lib/wsl/lib/libdxcore.so
```

I will try `release-1.12` branch later

**Comment by elezar** [comment]
@itmo153277 sorry, I meant once we have a fix pending. I will update you on this once we have more information.

**Comment by itmo153277** [comment]
Sure, I will test it

**Comment by elezar** [comment]
I have created https://gitlab.com/nvidia/container-toolkit/libnvidia-container/-/merge_requests/204 which may address this, but I will have to spend some time looking into this a bit more later. If that works, then we can backport that and release `1.12.2` later this week.

**Comment by itmo153277** [comment]
@elezar 
I tried your MR and it worked

```
$ nvidia-container-cli --version
cli-version: 1.12.1
lib-version: 1.12.1
build date: 2023-03-14T07:37+00:00
build revision: 7440a1ead8e4fe35edf1c973c73a662108b21a1f
build compiler: x86_64-linux-gnu-gcc-7 7.5.0
build platform: x86_64
build flags: -D_GNU_SOURCE -D_FORTIFY_SOURCE=2 -DNDEBUG -std=gnu11 -O2 -g -fdata-sections -ffunction-sections -fplan9-extensions -fstack-protector -fno-strict-aliasing -fvisibility=hidden -Wall -Wextra -Wcast-align -Wpointer-arith -Wmissing-prototypes -Wnonnull -Wwrite-strings -Wlogical-op -Wformat=2 -Wmissing-format-attribute -Winit-self -Wshadow -Wstrict-prototypes -Wunreachable-code -Wconversion -Wsign-conversion -Wno-unknown-warning-option -Wno-format-extra-args -Wno-gnu-alignof-expression -Wl,-zrelro -Wl,-znow -Wl,-zdefs -Wl,--gc-sections

$ docker run --gpus all nvcr.io/nvidia/k8s/cuda-sample:nbody nbody -gpu -benchmark
Run "nbody -benchmark [-numbodies=<numBodies>]" to measure performance.
        -fullscreen       (run n-body simulation in fullscreen mode)
        -fp64             (use double precision floating point values for simulation)
        -hostmem          (stores simulation data in host memory)
        -benchmark        (run benchmark to measure performance)
        -numbodies=<N>    (number of bodies (>= 1) to run in simulation)
        -device=<d>       (where d=0,1,2.... for the CUDA device to use)
        -numdevices=<i>   (where i=(number of CUDA devices > 0) to use for simulation)
        -compare          (compares simulation results running once on the default GPU and once on the CPU)
        -cpu              (run n-body simulation on the CPU)
        -tipsy=<file.bin> (load a tipsy model file for simulation)

NOTE: The CUDA Samples are not meant for performance measurements. Results may vary when GPU Boost is enabled.

> Windowed mode
> Simulation data stored in video memory
> Single precision floating point simulation
> 1 Devices used for simulation
GPU Device 0: "Ampere" with compute capability 8.6

> Compute 8.6 CUDA device: [NVIDIA GeForce RTX 3060 Laptop GPU]
30720 bodies, total time for 10 iterations: 31.059 ms
= 303.848 billion interactions per second
= 6076.951 single-precision GFLOP/s at 20 flops per interaction
```

**Comment by itmo153277** [comment]
Also `free(NULL)` is no-op, so the problem is probably `globfree` because `gl` is not initialized until much later. That would also mean that segfault is possible during `goto fail` in the other checks too.

**Comment by klueska** [comment]
Yes, you are right -- it's because we allocate the `gl` object on the stack so its internal `gl_pathv` variable is not necessarily `null`, causing this to error out:
```
https://github.com/lattera/glibc/blob/master/posix/globfree.c#LL30C14-L30C22
```

The correct fix is then to make sure that we `null` this out at the beginning, i.e.:
```
glob_t gl = {0};
```

**Comment by elezar** [comment]
Thanks @itmo153277 for the quick test. I have updated the MR as per @klueska's suggestion. If you get a chance to test that it would be great.

We will include this fix in our upcoming `1.13.0-rc.2` release and also backport it to the `1.12` branch for a `1.12.2` patch release.

**Comment by itmo153277** [comment]
@elezar I've already tried it, it works.

**Comment by elezar** [comment]
@itmo153277 we have included the fix in our `v1.13.0-rc.2` release that is now available from our experimental repositories. We will still be releasing a `v1.12.2` patch release by the end of next week with the fix.

**Comment by elezar** [comment]
@itmo153277 due to our `1.13.0` release we were not able to push this out as a patch release to the `v1.12.0` branch. Does the `v1.13.0` release that is now out resolve your issue?
